source/slang/syntax.cpp

```diff
@@ -940,6 +940,155 @@ void Type::accept(IValVisitor* visitor, void* extra)
         return this;
     }
 
+    // AssocTypeDeclRefType
+
+    String AssocTypeDeclRefType::ToString()
+    {
+        // TODO: what is appropriate here?
+        return "<AssocType>";
+    }
+
+    bool AssocTypeDeclRefType::EqualsImpl(Type * type)
+    {
+        if (auto assocTypeDeclRefType = type->As<AssocTypeDeclRefType>())
+        {
+            return declRef.Equals(assocTypeDeclRefType->declRef);
+        }
+        return false;
+    }
+
+    RefPtr<Val> AssocTypeDeclRefType::SubstituteImpl(Substitutions* subst, int* ioDiff)
+    {
+        if (!sourceType)
+            return this;
+        if (auto parentDeclRefType = sourceType->As<DeclRefType>())
+        {
```

Comment:
Okay, so here we have substitution logic for associated types, special-cased into `AssocTypeDeclRefType`, when it seems like it should really just apply to *any* `DeclRefType` that happens to reference an `AssocTypeDecl`.